pkg/cli/roundtripper.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strings"
+)
+
+type retryingRoundTripper struct {
+	wrapped http.RoundTripper
+}
+
+func (rt *retryingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return rt.roundTrip(req, 1, nil)
+}
+
+func (rt *retryingRoundTripper) roundTrip(req *http.Request, retry int, prevResp *http.Response) (*http.Response, error) {
+	if retry < 0 {
+		return prevResp, nil
+	}
+	resp, rtErr := rt.wrapped.RoundTrip(req)
+	if rtErr != nil {
+		return resp, rtErr
+	}
+	if resp.Header.Get("content-type") != "application/json" {
+		return resp, rtErr
+	}
+	b, err := io.ReadAll(resp.Body)
```

Comment:
@cenkalti is there anything the http headers that would indicate that we are dealing with an error? (so we can limit reading the full body only to the error scenarios?)

Reply:
Unfortunately no. The headers sent by kube-apiserver are as below:
```
HEADER Cache-Control no-cache, private
HEADER Content-Type application/json
HEADER X-Kubernetes-Pf-Flowschema-Uid 731b79e0-c1d1-4d8d-bba8-7588c969de79
HEADER X-Kubernetes-Pf-Prioritylevel-Uid 251f7bf0-8d47-4e8e-b644-6077a1fe19ab
HEADER Content-Length 119
HEADER Date Tue, 11 Oct 2022 13:17:31 GMT
HEADER Audit-Id 15493c2a-d08a-4410-b634-7b1302707bf5
```